### DISCUSSION TITLE
Can MOOSE users easily use SEACAS?

### DISCUSSION THREAD

**Discussion by WilkAndy** [idaholab/moose#16344] [Q&A Meshing] [answered]
libmesh is a submodule of MOOSE, and i see libmesh uses a lot of the libraries that SEACAS, https://github.com/gsjaardema/seacas , uses.    I don't know, but perhaps the entire SEACAS is hiding somewhere in libmesh.

I tried, and failed, to get SEACAS working on my computer - you have to get netcdf, hdf, and various other things, and i just ran out of time.

Does anyone have tips for using libmesh's already-built libraries to quickly get a working copy of SEACAS ?    (Or, you could suggest a well-documented alternative for writing exodus files using python.  The VTK libraries are almost perfect in this regard, but their exodusII writer has some bugs, boohoo.)

**Comment by milljm** [comment]
I am not sure how helpful this will be... But this is how I was building/including SEACAS tools into the old moose-environment package: https://github.com/idaholab/package_builder/blob/master/build_from_source/template/seacas-tools


The scripts therein might help you design your own installing method.

**Reply by WilkAndy** [reply]
Thanks @milljm

**Comment by cpgr** [comment]
Hey Andy, I had problems building SEACAS as well when using em2ex, especially on the Travis CI machines that I use for testing, so I wrote a pyexodus class that I use to write to an exodus file after converting from a Petrel file. It's in here: https://github.com/cpgr/em2ex/tree/master/pyexodus

It's pretty straightforward, and basically mimics the Exodus python API without needing SEACAS.

All it needs is the python netCDF4 which is easy to install

**Reply by WilkAndy** [reply]
Thanks @cpgr .  Do you know if libmesh already has netCDF4 ?

**Reply by cpgr** [reply]
I rely on the python one that is trivial to install `pip instal netCDF4`

I've also used `meshio` as @GregVernon mentioned and it is excellent, but didn't fit my particular use case (conversion from geological model).

**Comment by GregVernon** [comment, marked as answer]
@WilkAndy Here are several options for creating Exodus files via Python, depending on what your use-case is

## Translation from another format  
There's a fantastic Python module called [meshio](https://github.com/nschloe/meshio) (available on [PyPi](https://pypi.org/project/meshio/)) that can read, write, and convert between the following mesh formats:
> - Abaqus 
> - ANSYS msh
> - AVS-UCD
> -  CGNS
> -  DOLFIN XML
> -  Exodus 
> -  FLAC3D
> -  H5M
> -  Kratos/MDPA
> -  Medit
> -  MED/Salome
> -  Nastran (bulk data)
> -  Neuroglancer precomputed format
> -  Gmsh (format versions 2.2, 4.0, and 4.1)
> -  OBJ
> -  OFF
> -  PERMAS
> -  PLY
> -  STL
> -  Tecplot .dat
> -  TetGen .node/.ele
> -  SVG (2D only, output only)
> -  SU2
> -  UGRID
> -  VTK
> -  VTU
> -  WKT (TIN)
> -  XDMF

## Automated mesh-generation and export to Exodus
You could use Coreform Cubit, which has a Python-API that is also accessible as an importable Python module. and natively exports Exodus files.  I've automated many meshing routines (DoE, optimization, etc) using this technique in my former career as an FEA analyst, and as a student.  You can get a free edition of Coreform Cubit, called Coreform Cubit Learn, by following the [instructions on this page](https://coreform.com/products/coreform-cubit/free-meshing-software/).  The only limitations of the free edition are personal non-commercial use, educational use, and an export limitation of 50k elements.

## SEACAS
As noted by @cpgr, SEACAS includes a python module, `exodus.py` for directly querying or modifying Exodus files.  There are also some conversion tools that you might find helpful:
-  `exo2mat` and `mat2exo`  -- convert Exodus to / from Matlab file format
-  `exotxt` and `txtexo` -- convert Exodus to / from an equivalent ASCII format (which is documented somewhere)

**Reply by WilkAndy** [reply]
Excellent, thanks @GregVernon
